x-pack/legacy/plugins/apm/server/lib/alerting/error_occurrence/register_error_occurrence_alert_type.ts

```diff
@@ -0,0 +1,140 @@
+ */
+import { schema, TypeOf } from '@kbn/config-schema';
+import { InternalCoreSetup } from 'src/core/server';
+import { AggregationSearchResponse } from 'elasticsearch';
+import { idx } from '@kbn/elastic-idx';
+import { NOT_AVAILABLE_LABEL } from '../../../../common/i18n';
+import { APMError } from '../../../../typings/es_schemas/ui/APMError';
+import { ERROR_OCCURRENCE_ALERT_TYPE_ID } from '../../../../common/alerting/constants';
+import {
+  SERVICE_NAME,
+  PROCESSOR_EVENT,
+  ERROR_LOG_MESSAGE,
+  ERROR_EXC_MESSAGE,
+  ERROR_EXC_HANDLED,
+  ERROR_CULPRIT,
+  ERROR_GROUP_ID
+} from '../../../../common/elasticsearch_fieldnames';
+
+const paramsSchema = schema.object({
+  threshold: schema.number({ min: 1 }),
+  interval: schema.string(),
+  serviceName: schema.string()
+});
+
+const registerErrorOccurrenceAlertType = async (core: InternalCoreSetup) => {
+  const { alerting, elasticsearch } = core.http.server.plugins;
+
+  if (!alerting) {
+    throw new Error(
+      'Cannot register error occurrence alert type. Both the actions and alerting plugins need to be enabled'
+    );
+  }
+
+  alerting.registerType({
+    id: ERROR_OCCURRENCE_ALERT_TYPE_ID,
+    name: 'Error occurrences',
+    actionGroups: ['default'],
+    validate: {
+      params: paramsSchema
+    },
+    async executor({ params, services }) {
+      const { threshold, interval, serviceName } = params as TypeOf<
+        typeof paramsSchema
+      >;
+
+      const request = {
+        body: {
+          query: {
+            bool: {
+              filter: [
+                { term: { [SERVICE_NAME]: serviceName } },
+                { term: { [PROCESSOR_EVENT]: 'error' } },
+                {
+                  range: {
+                    '@timestamp': {
+                      gte: `now-${interval}`
+                    }
+                  }
+                }
+              ]
+            }
+          },
+          aggs: {
+            error_groups: {
+              terms: {
+                min_doc_count: threshold,
+                field: ERROR_GROUP_ID,
+                size: 10,
+                order: {
+                  _count: 'desc'
+                }
+              },
+              aggs: {
+                sample: {
+                  top_hits: {
+                    _source: [
+                      ERROR_LOG_MESSAGE,
+                      ERROR_EXC_MESSAGE,
+                      ERROR_EXC_HANDLED,
+                      ERROR_CULPRIT,
+                      ERROR_GROUP_ID,
+                      '@timestamp'
+                    ],
+                    sort: [
+                      {
+                        '@timestamp': 'desc'
+                      }
+                    ],
+                    size: 1
+                  }
+                }
+              }
+            }
+          }
+        }
+      };
+
+      const { callWithInternalUser } = elasticsearch.getCluster('admin');
```

Comment:
You can also use `services.callCluster(...)`. It will be in the context of the user who created the alert (security wise). The approach you have is fine as well if you don't want that.